mypy/checkexpr.py

```diff
@@ -3505,6 +3508,63 @@ def visit_op_expr(self, e: OpExpr) -> Type:
         else:
             raise RuntimeError(f"Unknown operator {e.op}")
 
+    def literal_value_from_expr(
+        self, expr: Expression, typ: Type | None = None
+    ) -> tuple[list[str | int], str, bool] | None:
+        if isinstance(expr, StrExpr):
+            return [expr.value], "builtins.str", False
+        if isinstance(expr, IntExpr):
+            return [expr.value], "builtins.int", False
+        if isinstance(expr, BytesExpr):
+            return [expr.value], "builtins.bytes", False
+
+        typ = typ or self.accept(expr)
+        ptype = get_proper_type(typ)
+
+        if isinstance(ptype, LiteralType) and not isinstance(ptype.value, (bool, float)):
+            return [ptype.value], ptype.fallback.type.fullname, True
+
+        if isinstance(ptype, UnionType):
+            fallback: str | None = None
+            values: list[str | int] = []
+            for item in ptype.items:
+                pitem = get_proper_type(item)
+                if not isinstance(pitem, LiteralType) or isinstance(pitem.value, (float, bool)):
+                    break
+                if fallback is None:
+                    fallback = pitem.fallback.type.fullname
+                if fallback != pitem.fallback.type.fullname:
+                    break
+                values.append(pitem.value)
+            else:
+                assert fallback is not None
+                return values, fallback, True
+        return None
+
+    def literal_expression_addition(self, e: OpExpr, left_type: Type) -> Type | None:
+        """Check if literal values can be combined with addition."""
+        assert e.op == "+"
+        if not (lvalue := self.literal_value_from_expr(e.left, left_type)):
+            return None
+        if (
+            not (rvalue := self.literal_value_from_expr(e.right))
+            or lvalue[1] != rvalue[1]  # different fallback
+            or lvalue[2] + rvalue[2] == 0  # no LiteralType
+        ):
+            return None
+
+        values: list[int | str] = sorted(
+            {
+                val[0] + val[1]  # type: ignore[operator]
```

Comment:
Can you avoid the `type: ignore` here? In the future we will probably disallow many ignores in code compiled with mypyc, since some type ignores can cause erratic behavior, such as C compilation errors.

Reply:
> Can you avoid the `type: ignore` here?

Not in an elegant way. The `values` could be either `str` or `int` (for int and bytes). With the comparison of the fallback types we know that left and right values have the same type but not which one. If necessary, it would be possible to split these into two cases with `cast` and set comprehensions for each one. Don't think that it would be an improvement though.